src/MultiRegion/cubed_sphere_grid.jl

```diff
@@ -401,16 +411,47 @@ function ConformalCubedSphereGrid(filepath::AbstractString,
     return MultiRegionGrid{FT, panel_topology..., CZ}(arch, partition, connectivity, region_grids, devices)
 end
 
-function with_halo(new_halo, csg::ConformalCubedSphereGrid)
-    region_rotation = []
+function with_halo(new_halo, csg::ConformalCubedSphereGrid{FT, TX, TY, TZ}) where {FT, TX, TY, TZ}
+    arch = csg.architecture
+    partition = csg.partition
+    connectivity = csg.connectivity
 
-    for region in 1:length(csg.partition)
+    region_rotation = []
+    for region in 1:length(partition)
         push!(region_rotation, csg[region].conformal_mapping.rotation)
     end
 
-    apply_regionally!(with_halo, new_halo, csg; rotation = Iterate(region_rotation))
+    region_grids = csg.region_grids
+    region_grids = construct_regionally(with_halo, new_halo, region_grids;
+                                        arch = CPU(), rotation = Iterate(region_rotation))
 
-    return csg
+    devices = Tuple(CPU() for _ in 1:length(partition))
+
+    CZ = typeof(getregion(region_grids, 1).z)
+
+    grid = MultiRegionGrid{FT, TX, TY, TZ, CZ}(CPU(),
+                                               partition,
+                                               connectivity,
+                                               region_grids,
+                                               devices)
+
+    fill_halo_regions!(grid)
+
+    # Now convert to user-prescribed architecture
+    region_grids = grid.region_grids
+    @apply_regionally new_region_grids = on_architecture(arch, region_grids)
+
+    new_devices = arch == CPU() ? Tuple(CPU() for _ in 1:length(partition)) : Tuple(CUDA.device() for _ in 1:length(partition))
```

Comment:
This will not allow you to run on multiple GPUs because `CUDA.device()` is pointing to the same GPU. Is there a reason why we cannot do
```suggestion
    new_devices = csg.region_grids.devices
```

Reply:
Thanks, @simone-silvestri. I see your point and you’re absolutely right. I was handling it in a long-winded way. I've now incorporated your suggestion in commit [39a13e2](https://github.com/CliMA/Oceananigans.jl/pull/4295/commits/39a13e2d8226b38b2ea3ad8cde3b3b342c43227a).